<a href="https://colab.research.google.com/github/SmrutiRanjan27/worldbank-global-education-analysis/blob/main/Smruti_Ranjan_WorldBank_Global_Education_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

The World Bank EdStats All Indicator Query holds over 4,000 internationally comparable indicators that describe education access, progression, completion, literacy, teachers, population, and expenditures.The indicators cover the education cycle from pre-primary to vocational and tertiary education and also holds learning outcome data from international and regional learning assessments (e.g. PISA, TIMSS, PIRLS), equity data from household surveys, and projection/attainment data.

Explore and analyze the data to identify variation of indicators across the globe, which countries are more alike and different. Feel free to add more extensive analyses and details.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd
import numpy as np

In [3]:
file_path = '/content/drive/MyDrive/AlmaBetter/Cohort Nilgiri/Module 1/Capstone Project/EdStatsData.csv'
df_stats = pd.read_csv(file_path)

## **Preliminary Analysis**

In [4]:
# lets look at basic info of the EdstatsData.csv
df_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 886930 entries, 0 to 886929
Data columns (total 70 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Country Name    886930 non-null  object 
 1   Country Code    886930 non-null  object 
 2   Indicator Name  886930 non-null  object 
 3   Indicator Code  886930 non-null  object 
 4   1970            72288 non-null   float64
 5   1971            35537 non-null   float64
 6   1972            35619 non-null   float64
 7   1973            35545 non-null   float64
 8   1974            35730 non-null   float64
 9   1975            87306 non-null   float64
 10  1976            37483 non-null   float64
 11  1977            37574 non-null   float64
 12  1978            37576 non-null   float64
 13  1979            36809 non-null   float64
 14  1980            89122 non-null   float64
 15  1981            38777 non-null   float64
 16  1982            37511 non-null   float64
 17  1983      

We have data for years starting from 2020 to 2100. Lets look at the details !

In [5]:
df_stats[['Country Name','Indicator Name','2085']][~df_stats['2085'].isna()]

Country Name                                     Indicator Name     2085
91309         World  Wittgenstein Projection: Mean years of schooli...     2.70
91310         World  Wittgenstein Projection: Mean years of schooli...     2.70
91311         World  Wittgenstein Projection: Mean years of schooli...     2.70
91312         World  Wittgenstein Projection: Mean years of schooli...    12.10
91313         World  Wittgenstein Projection: Mean Years of Schooli...     0.10
...             ...                                                ...      ...
886917     Zimbabwe  Wittgenstein Projection: Population in thousan...    91.70
886918     Zimbabwe  Wittgenstein Projection: Population in thousan...   205.47
886919     Zimbabwe  Wittgenstein Projection: Population in thousan...  3378.61
886920     Zimbabwe  Wittgenstein Projection: Population in thousan...  2924.73
886921     Zimbabwe  Wittgenstein Projection: Population in thousan...  6303.34

[51436 rows x 3 columns]

In [7]:
df_stats['Indicator Name'][~df_stats['2085'].isna()].unique().shape

(308,)

The columns with year values from 2020 to 2100 have non-null values which stores the information of preojected values of various Wittgenstein indicators

In [8]:
df_stats[df_stats['Country Name']=='Arab World']['Indicator Code']

0                 UIS.NERA.2
1               UIS.NERA.2.F
2             UIS.NERA.2.GPI
3               UIS.NERA.2.M
4                SE.PRM.TENR
                ...         
3660        UIS.LP.AG15T24.M
3661       SE.ADT.1524.LT.ZS
3662    SE.ADT.1524.LT.FE.ZS
3663    SE.ADT.1524.LT.FM.ZS
3664    SE.ADT.1524.LT.MA.ZS
Name: Indicator Code, Length: 3665, dtype: object

In [10]:
# lets look at the number of unique countries
df_stats['Country Name'].unique().shape

(242,)

In [11]:
# lets look at the number of unique indicators in the dataset
df_stats['Indicator Name'].unique().shape

(3665,)

In [12]:
#lets see if multiplying these two values gives us the total number of entries
3665*242

886930

**We have 242 countries and each of these countries have 3665 indicators !**

In [13]:
df_stats.head()

Country Name Country Code  ... 2100 Unnamed: 69
0   Arab World          ARB  ...  NaN         NaN
1   Arab World          ARB  ...  NaN         NaN
2   Arab World          ARB  ...  NaN         NaN
3   Arab World          ARB  ...  NaN         NaN
4   Arab World          ARB  ...  NaN         NaN

[5 rows x 70 columns]

In [14]:
# Just trying to get an idea of the dataset
df_stats[['Indicator Name','Indicator Code']][:10].values

array([['Adjusted net enrolment rate, lower secondary, both sexes (%)',
        'UIS.NERA.2'],
       ['Adjusted net enrolment rate, lower secondary, female (%)',
        'UIS.NERA.2.F'],
       ['Adjusted net enrolment rate, lower secondary, gender parity index (GPI)',
        'UIS.NERA.2.GPI'],
       ['Adjusted net enrolment rate, lower secondary, male (%)',
        'UIS.NERA.2.M'],
       ['Adjusted net enrolment rate, primary, both sexes (%)',
        'SE.PRM.TENR'],
       ['Adjusted net enrolment rate, primary, female (%)',
        'SE.PRM.TENR.FE'],
       ['Adjusted net enrolment rate, primary, gender parity index (GPI)',
        'UIS.NERA.1.GPI'],
       ['Adjusted net enrolment rate, primary, male (%)',
        'SE.PRM.TENR.MA'],
       ['Adjusted net enrolment rate, upper secondary, both sexes (%)',
        'UIS.NERA.3'],
       ['Adjusted net enrolment rate, upper secondary, female (%)',
        'UIS.NERA.3.F']], dtype=object)

**Lets look at the second dataset EdStatsCountry.csv**

In [15]:
file_path = '/content/drive/MyDrive/AlmaBetter/Cohort Nilgiri/Module 1/Capstone Project/EdStatsCountry.csv'
df_stats_country = pd.read_csv(file_path)

In [16]:
df_stats_country.head()

Country Code   Short Name  ... Latest water withdrawal data Unnamed: 31
0          ABW        Aruba  ...                          NaN         NaN
1          AFG  Afghanistan  ...                         2000         NaN
2          AGO       Angola  ...                         2005         NaN
3          ALB      Albania  ...                         2006         NaN
4          AND      Andorra  ...                          NaN         NaN

[5 rows x 32 columns]

In [17]:
# What are the different feature columns
df_stats_country.columns

Index(['Country Code', 'Short Name', 'Table Name', 'Long Name', '2-alpha code',
       'Currency Unit', 'Special Notes', 'Region', 'Income Group', 'WB-2 code',
       'National accounts base year', 'National accounts reference year',
       'SNA price valuation', 'Lending category', 'Other groups',
       'System of National Accounts', 'Alternative conversion factor',
       'PPP survey year', 'Balance of Payments Manual in use',
       'External debt Reporting status', 'System of trade',
       'Government Accounting concept', 'IMF data dissemination standard',
       'Latest population census', 'Latest household survey',
       'Source of most recent Income and expenditure data',
       'Vital registration complete', 'Latest agricultural census',
       'Latest industrial data', 'Latest trade data',
       'Latest water withdrawal data', 'Unnamed: 31'],
      dtype='object')

In [18]:
df_stats_country.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 241 entries, 0 to 240
Data columns (total 32 columns):
 #   Column                                             Non-Null Count  Dtype  
---  ------                                             --------------  -----  
 0   Country Code                                       241 non-null    object 
 1   Short Name                                         241 non-null    object 
 2   Table Name                                         241 non-null    object 
 3   Long Name                                          241 non-null    object 
 4   2-alpha code                                       238 non-null    object 
 5   Currency Unit                                      215 non-null    object 
 6   Special Notes                                      145 non-null    object 
 7   Region                                             214 non-null    object 
 8   Income Group                                       214 non-null    object 
 9   WB-2 code 

**We have economic stats for 241 of 242 countries in the EdStatsCountry dataset**

In [19]:
df_stats_country['Country Code'].unique()

array(['ABW', 'AFG', 'AGO', 'ALB', 'AND', 'ARB', 'ARE', 'ARG', 'ARM',
       'ASM', 'ATG', 'AUS', 'AUT', 'AZE', 'BDI', 'BEL', 'BEN', 'BFA',
       'BGD', 'BGR', 'BHR', 'BHS', 'BIH', 'BLR', 'BLZ', 'BMU', 'BOL',
       'BRA', 'BRB', 'BRN', 'BTN', 'BWA', 'CAF', 'CAN', 'CHE', 'CHI',
       'CHL', 'CHN', 'CIV', 'CMR', 'COD', 'COG', 'COL', 'COM', 'CPV',
       'CRI', 'CUB', 'CUW', 'CYM', 'CYP', 'CZE', 'DEU', 'DJI', 'DMA',
       'DNK', 'DOM', 'DZA', 'EAP', 'EAS', 'ECA', 'ECS', 'ECU', 'EGY',
       'EMU', 'ERI', 'ESP', 'EST', 'ETH', 'EUU', 'FIN', 'FJI', 'FRA',
       'FRO', 'FSM', 'GAB', 'GBR', 'GEO', 'GHA', 'GIB', 'GIN', 'GMB',
       'GNB', 'GNQ', 'GRC', 'GRD', 'GRL', 'GTM', 'GUM', 'GUY', 'HIC',
       'HKG', 'HND', 'HPC', 'HRV', 'HTI', 'HUN', 'IDN', 'IMN', 'IND',
       'IRL', 'IRN', 'IRQ', 'ISL', 'ISR', 'ITA', 'JAM', 'JOR', 'JPN',
       'KAZ', 'KEN', 'KGZ', 'KHM', 'KIR', 'KNA', 'KOR', 'KWT', 'LAC',
       'LAO', 'LBN', 'LBR', 'LBY', 'LCA', 'LCN', 'LDC', 'LIC', 'LIE',
       'LKA', 'LMC',

**Lets look at the third datset and get some info about EdStatsSeries !**

In [20]:
file_path = '/content/drive/MyDrive/AlmaBetter/Cohort Nilgiri/Module 1/Capstone Project/EdStatsSeries.csv'
df_series = pd.read_csv(file_path)

In [21]:
df_series.columns

Index(['Series Code', 'Topic', 'Indicator Name', 'Short definition',
       'Long definition', 'Unit of measure', 'Periodicity', 'Base Period',
       'Other notes', 'Aggregation method', 'Limitations and exceptions',
       'Notes from original source', 'General comments', 'Source',
       'Statistical concept and methodology', 'Development relevance',
       'Related source links', 'Other web links', 'Related indicators',
       'License Type', 'Unnamed: 20'],
      dtype='object')

In [22]:
df_series.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3665 entries, 0 to 3664
Data columns (total 21 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   Series Code                          3665 non-null   object 
 1   Topic                                3665 non-null   object 
 2   Indicator Name                       3665 non-null   object 
 3   Short definition                     2156 non-null   object 
 4   Long definition                      3665 non-null   object 
 5   Unit of measure                      0 non-null      float64
 6   Periodicity                          99 non-null     object 
 7   Base Period                          314 non-null    object 
 8   Other notes                          552 non-null    object 
 9   Aggregation method                   47 non-null     object 
 10  Limitations and exceptions           14 non-null     object 
 11  Notes from original source    

**So we have got detailed information about the each of 3665 indicators in EdStatsData in this dataset named EdstatsSeries.csv**

In [23]:
df_series.head()

Series Code       Topic  ... License Type Unnamed: 20
0  BAR.NOED.1519.FE.ZS  Attainment  ...          NaN         NaN
1     BAR.NOED.1519.ZS  Attainment  ...          NaN         NaN
2  BAR.NOED.15UP.FE.ZS  Attainment  ...          NaN         NaN
3     BAR.NOED.15UP.ZS  Attainment  ...          NaN         NaN
4  BAR.NOED.2024.FE.ZS  Attainment  ...          NaN         NaN

[5 rows x 21 columns]

In [24]:
# lets see if series code of EdStatsSeries is same as indicator code column of EdStatsData.csv
len(set(df_stats['Indicator Code'].unique())&set(df_series['Series Code']))

3612

**So we have detailed information of about 3612 of 3665 indicators in this dataset**

**I am curious about the periodicity and base period column, since we see a lot of NA values of a lot years for most of the indicators in EdStatsData dataset. This column might give us a clue about that !**

In [25]:
df_series[~df_series['Base Period'].isna()]['Base Period'].unique()

array(['2010', '2011', 'Projections (2010 to 2100)', '14'], dtype=object)

In [26]:
df_series[df_series['Base Period']=='Projections (2010 to 2100)'].shape[0]

308

**The number of indicators pertaining to only the projections data from 2010 to 2100 are 308 same as what we observed in EdStatsData.csv where we tried to see number unique indicators value for the years 2020-2100. Hence there are 308 rows for each country with projection data.**

In [27]:
df_series[~df_series['Periodicity'].isna()]['Periodicity'].unique()

array(['Annual'], dtype=object)

In [28]:
# Lets look at the series codes that are collected annually
annual_indicators = df_series[df_series['Periodicity'] == 'Annual']['Series Code'].values

In [29]:
for col in annual_indicators:
  print(df_stats[(df_stats['Country Name']=='World') & (df_stats['Indicator Code']==col)].values)

[['World' 'WLD' 'Personal computers (per 100 people)' 'IT.CMP.PCMP.P2'
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  1.99928817872655 2.2975139842551 2.4542773820406403 2.73917332505761
  3.0358366097334097 3.2701330051088 3.6227011073228996 4.14389535954102
  4.73463843841276 5.27217509808321 5.9537347095456505 6.80809200022134
  7.925872149644491 8.70785994978991 6.7591409852288695 7.9075438913605
  11.3897802152671 12.6751023333356 15.087313666953099 nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan]]
[['World' 'WLD' 'Internet users (per 100 people)' 'IT.NET.USER.P2' nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan 0.049687441232573 nan nan 0.25354299769265803 0.44997439569736297
  0.7823278288278471 1.33413785803189 2.0526629660289197 3.1563664453301
  4.653865215057491 6.770066508052451 8.095458176900848
  10.579787734322801 12.2752833522124 14.172

**Lets look at the fourth data in EdStatsCountry-Series.csv**

In [30]:
file_path = '/content/drive/MyDrive/AlmaBetter/Cohort Nilgiri/Module 1/Capstone Project/EdStatsCountry-Series.csv'
df_country_series = pd.read_csv(file_path)

In [31]:
df_country_series.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 613 entries, 0 to 612
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   CountryCode  613 non-null    object 
 1   SeriesCode   613 non-null    object 
 2   DESCRIPTION  613 non-null    object 
 3   Unnamed: 3   0 non-null      float64
dtypes: float64(1), object(3)
memory usage: 19.3+ KB


In [32]:
df_country_series.head()

CountryCode  ... Unnamed: 3
0         ABW  ...        NaN
1         ABW  ...        NaN
2         AFG  ...        NaN
3         AFG  ...        NaN
4         AFG  ...        NaN

[5 rows x 4 columns]

**Lets look at the fourth data in EdStatsSeries.csv**

In [33]:
file_path = '/content/drive/MyDrive/AlmaBetter/Cohort Nilgiri/Module 1/Capstone Project/EdStatsFootNote.csv'
df_footnote = pd.read_csv(file_path)

In [34]:
df_footnote.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 643638 entries, 0 to 643637
Data columns (total 5 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   CountryCode  643638 non-null  object 
 1   SeriesCode   643638 non-null  object 
 2   Year         643638 non-null  object 
 3   DESCRIPTION  643638 non-null  object 
 4   Unnamed: 4   0 non-null       float64
dtypes: float64(1), object(4)
memory usage: 24.6+ MB


In [35]:
df_footnote.head()

CountryCode      SeriesCode    Year          DESCRIPTION  Unnamed: 4
0         ABW  SE.PRE.ENRL.FE  YR2001  Country estimation.         NaN
1         ABW  SE.TER.TCHR.FE  YR2005  Country estimation.         NaN
2         ABW  SE.PRE.TCHR.FE  YR2000  Country estimation.         NaN
3         ABW  SE.SEC.ENRL.GC  YR2004  Country estimation.         NaN
4         ABW     SE.PRE.TCHR  YR2006  Country estimation.         NaN

### Preliminary Observations from all datasets :


1.   df_stats_country contains a list of all the 241 countries and their economic stats 
2.   df_stats_data contains data for combination of each country and 3665 indicators from 1970 to 2100
3.   df_series contains detailed information about most of the indicators in df_stats_data
4.   df_country_series contains some basic information about all the indicators estimated for certain countries
5.   df_footnote contains information about uncertaininty and origin of data series used for the indicators for each year






